### ISSUE TITLE
NullPointerException in printTestResults

### ISSUE THREAD

**Issue by FilenkovMaxim** [schroepf/TestLab#21]
Hello. 
When I tried to run tests I got this error and and not a single report file.

```
2) Test mechanism
java.lang.NullPointerException: Attempt to invoke virtual method 'void java.io.Writer.write(char[], int, int)' on a null object reference
	at org.kxml2.io.KXmlSerializer.flushBuffer(KXmlSerializer.java:80)
	at org.kxml2.io.KXmlSerializer.flush(KXmlSerializer.java:503)
	at org.kxml2.io.KXmlSerializer.endDocument(KXmlSerializer.java:203)
	at de.schroepf.androidxmlrunlistener.XmlRunListener.printTestResults(XmlRunListener.java:238)
	at de.schroepf.androidxmlrunlistener.XmlRunListener.testRunFinished(XmlRunListener.java:148)
	at org.junit.runner.notification.SynchronizedRunListener.testRunFinished(SynchronizedRunListener.java:42)
	at org.junit.runner.notification.RunNotifier$2.notifyListener(RunNotifier.java:103)
	at org.junit.runner.notification.RunNotifier$SafeNotifier.run(RunNotifier.java:72)
	at org.junit.runner.notification.RunNotifier.fireTestRunFinished(RunNotifier.java:100)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:138)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at android.support.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
	at android.support.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:384)
	at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:1932)
```

**Comment by schroepf** [comment]
Hi @FilenkovMaxim,

thanks for the bug report. Do you have more information on how to reproduces this issue? On which deivce/emulator were you running the tests and which android version did you use?

**Comment by ovitrif** [comment]
I have the same issue on a Pixel_2_XL_API_28

**Comment by schroepf** [comment]
so far I was not able to reproduce this issue

can you provide me some more details? which version of android-xml-run-listener do you use? does the issue still occur with an older version than the one you are using?

**Comment by ghost** [comment]
Similar issue for me
this is my gradle file [build.gradle](https://gist.github.com/codingotaku/700c805c5dbfff66c0874f0b4722a72a)

```
I/TestRunner: java.lang.NullPointerException
        at org.kxml2.io.KXmlSerializer.startTag(KXmlSerializer.java:406)
        at de.schroepf.androidxmlrunlistener.XmlRunListener.printTestResults(XmlRunListener.java:178)
        at de.schroepf.androidxmlrunlistener.XmlRunListener.testRunFinished(XmlRunListener.java:148)
        at org.junit.runner.notification.SynchronizedRunListener.testRunFinished(SynchronizedRunListener.java:42)
        at org.junit.runner.notification.RunNotifier$2.notifyListener(RunNotifier.java:103)
        at org.junit.runner.notification.RunNotifier$SafeNotifier.run(RunNotifier.java:72)
        at org.junit.runner.notification.RunNotifier.fireTestRunFinished(RunNotifier.java:100)
        at org.junit.runner.JUnitCore.run(JUnitCore.java:138)
        at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
        at android.support.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
        at android.support.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:384)
        at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:1701)
```

**Comment by schroepf** [comment]
Hi @codingotaku, @ovitrif, @FilenkovMaxim,

I think I was able to reproduce the issue: when using an app module which hasn't been migrated to AndroidX yet together with `android-xml-run-listener:0.3.1` which has already been migrated to AndroidX...

did all 3 of you use a similar setup (i.e. android-xml-run-listener:0.3.1 used from a module without anroidX support)?

can you please try if going back to version 0.3.0 of android-xml-run-listener fixes the problem for you?

**Comment by ovitrif** [comment]
In my case it's true there has been no migration to AndroidX on my project. I'll check more tomorrow

**Comment by ghost** [comment]
> did all 3 of you use a similar setup (i.e. android-xml-run-listener:0.3.1 used from a module without anroidX support)?
> 
> can you please try if going back to version 0.3.0 of android-xml-run-listener fixes the problem for you?

yes, that seems to be the problem, I fixed it in the code after cloning your repository, everything seems to be working fine as of now.

**Comment by schroepf** [comment]
Ok, thanks!

looks like the reason for this bug is that my custom `InstrumentationRunListener` implementation cannot access the `Instrumentation` (and thus the `Context` to store the xml reports) as long as the listener already uses androidX and the app at the same time installs it's dependencies the "old" way (i.e. not from androidx)

so far I couldn't find a fix for this issue, so I'm thinking about providing two libraries in the future, something like:
- android-xml-run-listener: for apps that use the "old" way
- androidx-xml-run-listener: for apps that use AndroidX

in the mean time it should be ok to just use the 0.3.0 version of  android-xml-run-listener instead of 0.3.1 (as androidx support is the only difference between the two versions)

**Comment by ovitrif** [comment]
@schroepf that might be a solution, indeed

**Comment by schroepf** [comment]
Finally, today I found some time to create a dedicated library build for non-AndroidX projects.
As soon as it's published to jcenter the "legacy" build will be available as: `de.schroepf:android-xml-run-listener-legacy:0.4.0`, the sources for this build are maintained in the `legacy/android-xml-run-listener` branch of this git repository.

As I expect the AndroidX compatible version of this library to be around longer than the legacy version I decided not to change it's artifact name and remain with the original name of this library.

PR #22 updates the `README` and updated version to 0.4.0 to be consistent with the legacy build version. As soon as the legacy artifact was published to jcenter I will merge PR #22 and close this issue.

If you still run into issues with the legacy build please let me know (reopen this issue or create a new one).

**Comment by schroepf** [comment]
closing issue. PR #22 is merged.
- AndroidX users please continue to use `de.schroepf:android-xml-run-listener:0.4.0`
- non andoridX users please use the new `de.schroepf:android-xml-run-listener-legacy:0.4.0`
